lib/Sema/CSSimplify.cpp

```diff
@@ -9693,14 +9712,22 @@ ConstraintSystem::addKeyPathApplicationRootConstraint(Type root, ConstraintLocat
           path[0].getKind() == ConstraintLocator::SubscriptMember) ||
          (path.size() == 2 &&
           path[1].getKind() == ConstraintLocator::KeyPathDynamicMember));
+
   auto indexTuple = dyn_cast<TupleExpr>(subscript->getIndex());
-  if (!indexTuple || indexTuple->getNumElements() != 1)
+  auto indexParen = dyn_cast<ParenExpr>(subscript->getIndex());
+  // If a keypath subscript is used without the expected `keyPath:` label,
+  // continue with type-checking when attempting fixes so that it gets caught
+  // by the argument label checking. In such cases, the KeyPathExpr is contained
+  // in a ParenExpr, instead of a TupleExpr.
+  if ((!indexTuple || indexTuple->getNumElements() != 1) &&
+      !(indexParen && shouldAttemptFixes()))
```

Comment:
I think we can turn this into an assert which check for argument being either a single element tuple or a paren. I think that's better than simply return here, can `dyn_cast` in `KeyPathExpr` could be switched into `cast`.

Reply:
I agree on the assert, but it does not seem that we can turn the `KeyPathExpr` `dyn_cast` into a `cast`. Doing so causes crashes when building the standard library.

Reply:
Now that's interesting, but it's okay to address that separately.